src/qz/printer/info/NativePrinterList.java

```diff
@@ -0,0 +1,81 @@
+package qz.printer.info;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import qz.utils.SystemUtilities;
+
+import javax.print.PrintService;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public abstract class NativePrinterList extends ConcurrentHashMap<String, NativePrinter> {
+    private static NativePrinterList instance;
+
+    public abstract NativePrinterList putAll(PrintService[] services);
+    abstract void fillAttributes(NativePrinter printer);
+
+    public static NativePrinterList getInstance() {
+        if (instance == null) {
+            if (SystemUtilities.isWindows()) {
+                instance = new WindowsPrinterList();
+            } else {
+                instance = new CupsPrinterList();
+            }
+        }
+        return instance;
+    }
+    private static final Logger log = LoggerFactory.getLogger(NativePrinterList.class);
+
+    public String getPrinterId(String description) {
+        for(Map.Entry<String,NativePrinter> entry : entrySet()) {
+            NativePrinter info = entry.getValue();
+            if (info.getDescription().equals(description)) {
+                return entry.getKey();
+            }
+        }
+        log.warn("Could not find printerId for " + description);
+        return description;
+    }
+
+    public PrintService[] findMissing(PrintService[] services) {
+        ArrayList<PrintService> toAdd = new ArrayList<>();
+        // Flag outdated
+        for (NativePrinter printer : values()) {
+            printer.setOutdated(true);
+        }
+        // Unflag
+        Collection<NativePrinter> shrinkingList = new ArrayList<>(values()); // shrinking list drastically improves performance
+        boolean found = false;
+        for (PrintService service : services) {
```

Comment:
Done, but this requires us to trust the pointer, otherwise the internal ArrayList `contains(...)`, calls the underlying PrintService's `equals(...)`, which gets back into calling `.getName(...)` over and over, causing 700ms delays.

The silver lining is if we trust the pointer reference, we can get the loop down to 0ms, which is a huge improvement.  Code available for review here: https://github.com/qzind/tray/commit/8c58c8df904307b6ec62a96610bece1bfc99a668#diff-03936d0d852ac8c772a90f88d9bdda8fR58

Reply:
The code looks alright, though I'd like to get it tested in a few different scenarios (ie. multi-connections and reconnects) to make sure the pointer references won't be a problem.

Reply:
This new technique even seems resilient to killing the CUPS daemon and staring it back up.

```bash
sudo launchctl unload /System/Library/LaunchDaemons/org.cups.cupsd.plist
# 0 printers found
```

```bash
sudo launchctl load /System/Library/LaunchDaemons/org.cups.cupsd.plist
9 printers found.
```

Java even reuses the pointers internally from before the CUPS shutdown.  I think this PR is in good shape.